packages/inter-protocol/src/auction/auctioneer.js

```diff
@@ -0,0 +1,282 @@
+import '@agoric/zoe/exported.js';
+import '@agoric/zoe/src/contracts/exported.js';
+import '@agoric/governance/exported.js';
+
+import { Far } from '@endo/marshal';
+import { E } from '@endo/eventual-send';
+import {
+  M,
+  makeScalarBigMapStore,
+  provide,
+  provideDurableMapStore,
+} from '@agoric/vat-data';
+import { AmountMath } from '@agoric/ertp';
+import {
+  atomicRearrange,
+  makeRatioFromAmounts,
+  makeRatio,
+  natSafeMath,
+  floorMultiplyBy,
+  provideEmptySeat,
+} from '@agoric/zoe/src/contractSupport/index.js';
+import { handleParamGovernance } from '@agoric/governance';
+import { makeTracer } from '@agoric/internal';
+import { FullProposalShape } from '@agoric/zoe/src/typeGuards.js';
+
+import { makeAuctionBook } from './auctionBook.js';
+import { BASIS_POINTS } from './util.js';
+import { makeScheduler } from './scheduler.js';
+import { auctioneerParamTypes } from './params.js';
+
+/** @typedef {import('@agoric/vat-data').Baggage} Baggage */
+
+const { Fail, quote: q } = assert;
+
+const trace = makeTracer('Auction', false);
+
+const makeBPRatio = (rate, currencyBrand, collateralBrand = currencyBrand) =>
+  makeRatioFromAmounts(
+    AmountMath.make(currencyBrand, rate),
+    AmountMath.make(collateralBrand, 10000n),
+  );
+
+/**
+ * @param {ZCF<GovernanceTerms<{
+ *   StartFrequency: 'relativeTime',
+ *   ClockStep: 'relativeTime',
+ *   StartingRate: 'nat',
+ *   lowestRate: 'nat',
+ *   DiscountStep: 'nat',
+ * }> & {timerService: import('@agoric/time/src/types').TimerService, priceAuthority: PriceAuthority}>} zcf
+ * @param {{initialPoserInvitation: Invitation, storageNode: StorageNode, marshaller: Marshaller}} privateArgs
+ * @param {Baggage} baggage
+ */
```

Comment:
In order to review this, I found myself reformatting it to

```suggestion
/**
 * @param {ZCF<GovernanceTerms<{
 *   StartFrequency: 'relativeTime',
 *   ClockStep: 'relativeTime',
 *   StartingRate: 'nat',
 *   lowestRate: 'nat',
 *   DiscountStep: 'nat',
 * }> & {
 *   timerService: import('@agoric/time/src/types').TimerService,
 *   priceAuthority: PriceAuthority
 * }>} zcf
 * @param {{
 *   initialPoserInvitation: Invitation,
 *   storageNode: StorageNode,
 *   marshaller: Marshaller
 * }} privateArgs
 * @param {Baggage} baggage
 */
```

Reply:
The params to `GovernedTerms` seems annoyingly redundant with `auctioneerParamTypes`. Can we reduce the redundancy? If not, or in the meantime, are they checked against each other for consistency?

Reply:
Aside from `auctioneerParamTypes`, the `privateArgs` are exactly what is needed for `handleParamGovernance`. Is this expected to be a typical pattern? Any way to make this more convenient?

Reply:
I agree on the redundancy.  I don't think they're checked directly against each other, but I think typescript knows that the params returned from `handleParamGovernance` (line 157) match this declaration, and compare that to the requirements of makeScheduler, which declares `params` to be `AuctionParamManaager` which was declared in `params.js`.

> `privateArgs` are exactly what is needed for `handleParamGovernance`. Is this expected to be a typical pattern?
yes.

@turadg understands these interconnections much better than I do.  Would you respond?

Reply:
> Any way to make this more convenient?

There's a bunch of refactoring we could do to governance. I'd like to make it require less repetition. Not a priority tho.

For this particular case we could have a type like `ParamGovernancePrivateArgs` and union that here. But the args `marshaller` and `storageNode` probably will leave `handleParamGovernance` once we stop using the deprecated StoredPublishKit for governance. I expect that to happen when we [make governance durable](https://github.com/Agoric/agoric-sdk/issues/5200).